File: Python/mazo.py
Oros = "Oros"
Espadas = "Espadas"
Basto = "Basto"
Copas = "Copas"
Cualquier = "Cualquier"

def mismoPalo(a,b):
    if a == Cualquier or b == Cualquier : return True
    else : return a == b

As = 1
Sota = 10
Caballo = 11
Rey = 12

def Carta(n, p):
    if n in [*(range(As,8)),*(range(Sota,Rey+1))] and p in [Oros,Espadas,Basto,Copas, Cualquier] : return {'numero' : n, 'palo' : p}
    else : return {'numero' : 0, 'palo' : Cualquier}

Macho = Carta(As,Espadas)
Hembra = Carta(As, Basto)
AnchoOro = Carta(As, Oros)
AnchoCopa = Carta(As, Copas)
Nada = Carta(0,Cualquier)

def mostrarCarta(c):
    if c == Macho : return "Macho"
    elif c == Hembra : return "Hembra"
    elif c == AnchoOro : return "Ancho de Oros"
    elif c == AnchoCopa : return "Ancho de Copas"
    elif c == Nada : return ""
    elif c['numero'] == As : return "As de " + c['palo']
    elif c['numero'] == Sota : return "Sota de " + c['palo']
    elif c['numero'] == Caballo : return "Caballo de " + c['palo']
    elif c['numero'] == Rey : return "Rey de " + c['palo']
    else : return str(c['numero']) + " de " + c['palo']

Orden = [Macho, Hembra, Carta(7,Espadas), Carta(7,Oros), Carta(3,Cualquier), Carta(2,Cualquier),
        Carta(Rey,Cualquier), Carta(Caballo,Cualquier), Carta(7,Cualquier), Carta(6,Cualquier),
        Carta(5,Cualquier), Carta(4,Cualquier), Nada]

Mazo = [Carta(x,y) for x in [*(range(As,8)),*(range(Sota,Rey+1))] for y in [Oros,Espadas,Basto,Copas]]

def cartasIguales(a,b):
    return a['numero'] == b['numero'] and cartaMismoPalo(a,b)

def cartaEsMayor(a,b):
    i1 = 0
    i2 = 0
    for carta in Orden :
        if cartasIguales(a,carta) : i1 = Orden.index(carta)
        if cartasIguales(b,carta) : i2 = Orden.index(carta)
    return i1 < i2

def valorPEnvido(c):
    if c['numero'] < Sota : return c['numero']
    else : return 0

def cartaMismoPalo(c1,c2):
    return mismoPalo(c1['palo'],c2['palo'])

File: Haskell/Mazo.hs
module Mazo
(
    Carta(..),
    Palo(..),
    carta,
    mazo,
    as,
    rey,
    sota,
    caballo,
    macho,
    hembra,
    anchoCopa,
    anchoOro,
    nada,
    orden,
    paloIgual,
    valorPEnvido
) where

import Data.List( elemIndex )

-- Definimos los palos
data Palo = Oros | Espadas | Basto | Copas | Cualquier deriving ( Show )

-- Definimos la igualdad de los palos
instance Eq Palo where
    ( == ) Cualquier _ = True
    ( == ) _ Cualquier = True
    ( == ) Oros Oros = True 
    ( == ) Espadas Espadas = True
    ( == ) Basto Basto = True 
    ( == ) Copas Copas = True
    ( == ) _ _ = False

---------------------------------------------------------
-- Definimos los nombres comunes para las cartas con cara
as :: Int
as = 1

sota :: Int
sota = 10

caballo :: Int
caballo = 11

rey :: Int
rey = 12
---------------------------------------------------------

-- Definimos lo que es una carta
data Carta = C { numero :: Int, palo :: Palo}

------------------------------------------------------------------------------------------------------------
-- Cuando una carta es mayor que otra
instance Ord Carta where
    compare c1 c2 = compare (elemIndex c2 orden) (elemIndex c1 orden)

-- Cuando una carta es igual a otra 
instance Eq Carta where
    (==) c1 c2 
        | numero c1 == numero c2 = paloIgual c1 c2
        | otherwise = False
-------------------------------------------------------------------------------------------------------------

-- Definimos cómo se imprime una carta por pantalla
instance Show Carta where
    show c
        | c == macho = "Macho"
        | c == hembra = "Hembra"
        | c == anchoCopa = "Ancho de Copas"
        | c == anchoOro = "Ancho de Oros"
        | c == nada = ""
        | numero c == as = "As de " ++ show (palo c)
        | numero c == sota = "Sota de " ++ show (palo c)
        | numero c == caballo = "Caballo de " ++ show (palo c)
        | numero c == rey = "Rey de " ++ show (palo c)
        | otherwise = show (numero c) ++ " de " ++ show (palo c)

--------------------------------------------------------
-- Definimos los nombres de las cartas importantes
macho :: Carta
macho = C as Espadas

hembra :: Carta
hembra = C as Basto

anchoOro :: Carta
anchoOro = C as Oros

anchoCopa :: Carta
anchoCopa = C as Copas
--------------------------------------------------------

-- Definimos el orden específico de importancia para las cartas en el truco, ésto facilita la comparación
orden :: [Carta]
orden = [macho, hembra, C 7 Espadas, C 7 Oros, C 3 Cualquier , C 2 Cualquier , C 1 Cualquier ,
        C rey Cualquier, C caballo Cualquier, C 7 Cualquier, C 6 Cualquier, C 5 Cualquier, C 4 Cualquier, nada]

nada :: Carta
nada = C 0 Cualquier

-- Listamos TODAS las cartas que se usan en el truco
mazo :: [Carta]
mazo = [ C v p | v <- [as..7]++[sota..rey], p <- [Oros, Espadas, Basto, Copas]]

-- Construir una carta en base a un número y un palo. Validamos contra las cartas del mazo
carta :: Int -> Palo -> Carta
carta v p
    | v `elem` [as..7]++[sota..rey] = C v p
    | otherwise = nada

-- Cuánto vale cada carta al momento de sumar su valor para el canto de envido
valorPEnvido :: Carta -> Int
valorPEnvido c
    | numero c < sota = numero c
    | otherwise = 0

-- Comparar si dos cartas son del mismo palo. (Para simplificar notación)
paloIgual :: Carta -> Carta -> Bool 
paloIgual c1 c2 = palo c1 == palo c2
